### PR TITLE
Add support for JWT authentication in Stargate

### LINKED ISSUE #531: K8SSAND-1473 ⁃ StargateSpec configuration for JWT authorization with external authentication
**What is missing?**
I am in great need of JWT authorization with external authentication, and I have voiced this earlier in 2021 in the K8ssandra repository, https://github.com/k8ssandra/k8ssandra/issues/1158. K8ssandra operator also doesn't [expose JAVA_OPTS](https://github.com/k8ssandra/k8ssandra/issues/1179), which would be a fine workaround for now. Moreover, it is a very crude implementation, and these fields should be exposed natively to the manifest.

I think @dougwettlaufer is more educated than me on this topic. Refer to this [post](https://github.com/stargate/docs/issues/201#issuecomment-1086776419) to learn about requirements for JWT authorization with external authentication on Stargate; semi-verified [example](https://github.com/stargate/docs/issues/201#issuecomment-1108851023). Fields under `JAVA_OPTS` that should be exposed:
- `-D stargate.auth_id` (this will be set to `AuthJwtService` in my case)
- `-D stargate.auth.jwt_provider_url`, JWKS URL
- ~`-D stargate.cql_use_transitional_auth` (`true` when authentication is upstream)~

**Why do we need it?**
This is a natively supported feature by Stargate, and it should be supported along with other native Stargate features. JWT has advantages over password authentication, the main benefit is the reduced load on the IDP -> lowered latency -> faster uploads.

**Environment**

* K8ssandra Operator version:

```
Name:                   k8ssandra-operator
Namespace:              k8ssandra
CreationTimestamp:      Sat, 23 Apr 2022 21:56:46 +0200
Labels:                 app.kubernetes.io/instance=k8ssandra-operator
app.kubernetes.io/managed-by=Helm
app.kubernetes.io/name=k8ssandra-operator
app.kubernetes.io/part-of=k8ssandra-k8ssandra-operator-k8ssandra
control-plane=k8ssandra-operator
helm.sh/chart=k8ssandra-operator-0.37.1
Annotations:            deployment.kubernetes.io/revision: 1
meta.helm.sh/release-name: k8ssandra-operator
meta.helm.sh/release-namespace: k8ssandra
Selector:               app.kubernetes.io/instance=k8ssandra-operator,app.kubernetes.io/name=k8ssandra-operator,app.kubernetes.io/part-of=k8ssandra-k8ssandra-operator-k8ssandra,control-plane=k8ssandra-operator
Replicas:               1 desired | 1 updated | 1 total | 1 available | 0 unavailable
StrategyType:           RollingUpdate
MinReadySeconds:        0
RollingUpdateStrategy:  25% max unavailable, 25% max surge
Pod Template:
Labels:           app.kubernetes.io/instance=k8ssandra-operator
app.kubernetes.io/managed-by=Helm
app.kubernetes.io/name=k8ssandra-operator
app.kubernetes.io/part-of=k8ssandra-k8ssandra-operator-k8ssandra
control-plane=k8ssandra-operator
helm.sh/chart=k8ssandra-operator-0.37.1
Service Account:  k8ssandra-operator
Containers:
k8ssandra-operator:
Image:      docker.io/k8ssandra/k8ssandra-operator:v1.0.1
Port:       9443/TCP
Host Port:  0/TCP
Command:
/manager
Liveness:   http-get http://:8081/healthz delay=15s timeout=1s period=20s #success=1 #failure=3
Readiness:  http-get http://:8081/readyz delay=5s timeout=1s period=10s #success=1 #failure=3
Environment:
WATCH_NAMESPACE:           (v1:metadata.namespace)
K8SSANDRA_CONTROL_PLANE:  true
Mounts:
/controller_manager_config.yaml from manager-config (rw,path="controller_manager_config.yaml")
/tmp/k8s-webhook-server/serving-certs from cert (ro)
Volumes:
manager-config:
Type:      ConfigMap (a volume populated by a ConfigMap)
Name:      k8ssandra-operator-manager-config
Optional:  false
cert:
Type:        Secret (a volume populated by a Secret)
SecretName:  k8ssandra-operator-webhook-server-cert
Optional:    false
Conditions:
Type           Status  Reason
----           ------  ------
Available      True    MinimumReplicasAvailable
Progressing    True    NewReplicaSetAvailable
OldReplicaSets:  <none>
NewReplicaSet:   k8ssandra-operator-6894cf5cdd (1/1 replicas created)
Events:
Type    Reason             Age   From                   Message
----    ------             ----  ----                   -------
Normal  ScalingReplicaSet  32m   deployment-controller  Scaled up replica set k8ssandra-operator-6894cf5cdd to 1
```

**Anything else we need to know?**:
Academic project

┆Issue is synchronized with this [Jira Task](https://k8ssandra.atlassian.net/browse/K8SSAND-1473) by [Unito](https://www.unito.io)
┆friendlyId: K8SSAND-1473
┆priority: Medium

### PR DESCRIPTION
**Pull request by adutra**
**What this PR does**:
Adds support for JWT authentication in Stargate.

**Which issue(s) this PR fixes**:
Fixes #531.

**Checklist**
- [x] Changes manually tested
- [x] Automated Tests added/updated
- [ ] Documentation added/updated
- [x] CHANGELOG.md updated (not required for documentation PRs)
- [x] CLA Signed:  [DataStax CLA](https://cla.datastax.com/)

### CODE REVIEW

**Comment by adutra** [comment]
I was finally able to add an e2e test. Ready for review.

**Comment by adutra** [comment]
@caniko would you be able to test this PR with your own setup? Thanks!

**Comment by caniko** [comment]
> @caniko would you be able to test this PR with your own setup? Thanks!

@adutra, I have been busy with an urgent analysis project. I will test this week. Thank you!

File: pkg/stargate/deployments.go
Comment (adejanovski) [inline pkg/stargate/deployments.go:441]: Awesome comment!

**Review by adejanovski — APPROVED** [review]
LGTM, with just a minor nit on some repeated variables.

**Comment by adutra** [comment]
I've rebased and fixed the conflicts. Waiting for the CI runs to finish.

**Comment by adutra** [comment]
It seems the UpgradeOperatorImage e2e test is broken. I will look into it before merging.

**Comment by adejanovski** [comment]
> It seems the UpgradeOperatorImage e2e test is broken. I will look into it before merging.

You may want to rebase on top of main. I fixed it there as there were some timing issues with Stargate upgrades.

File: apis/stargate/v1alpha1/stargate_types.go
Comment (adejanovski) [reply apis/stargate/v1alpha1/stargate_types.go:191]: It is, and it seems like it's breaking our upgrade e2e test.

**Review by adejanovski — CHANGES_REQUESTED** [review]
As pointed out by @jsanda, the change to the Stargate CRD is a breaking one. 
This is what's causing trouble in the upgrade test because the CRDs loaded are the newest but the k8ssandra-operator image is 1.0.0, which then prevents the creation of the Stargate CR with `auth: true`. While this is incorrect and should be done differently, I assume folks would run into troubles during upgrades anyway since there's no converting webhook.
We need to figure out if we want to keep the old auth  setting on top of the newly introduced structure and deal with those in the controller, or bump the api version and add a converting webhook.

File: apis/stargate/v1alpha1/stargate_types.go
Comment (adutra) [reply apis/stargate/v1alpha1/stargate_types.go]: I didn't know we cared about breaking changes in v1alpha1. There might be many other breaking changes already between 1.0 and 1.2.

Comment (adejanovski) [reply apis/stargate/v1alpha1/stargate_types.go]: I think we should adhere to a policy where breaking changes are supported from one version to the next, using a converting webhook.
As unstable as our APIs can be for now, upgrades should always be possible.
If we follow the [Kubernetes versioning policies](https://kubernetes.io/docs/reference/using-api/deprecation-policy/#deprecating-parts-of-the-api), having alpha level APIs mean we can remove them without any prior deprecation, but in our case the problem is slightly different: K8ssandraOperator generates the Stargate resources and should be able to handle any Stargate API change gracefully, including the upgrades.

**Comment by adutra** [comment]
I reinstated the boolean field. As usual with Kubernetes, the change itself took 5 minutes, and local re-testing took 6 hours. But I expect all tests to pass now. Only for some strange reason, they didn't run on my last commit. I'll kick a manual run.

File: test/e2e/suite_test.go
Comment (adutra) [reply test/e2e/suite_test.go:1602]: Sorry I forgot to reply: I don't see the value of doing so. If the code is expressive enough, extracting constants or global variables don't necessarily improve things. Besides, we have the `polling` struct, so I'd say that the values should be extracted there, as new fields. But again, unless we need different timeouts per environment – which isn't the case here – creating a new field in `polling` is not necessarily more expressive or clean than defining the timeout next to where it is going to be used.

Comment (adejanovski) [reply test/e2e/suite_test.go:1602]: 👌

**Review by adejanovski — APPROVED** [review]

**Pull request merged**
